### ISSUE TITLE
guix-edit does not always honour GUIX_PACKAGE_PATH

### ISSUE THREAD

**Issue by jellelicht** [alezost/guix.el#1]
It seems that `guix-edit` and related locations functionality are not able to understand external guix package definitions. Whereas the actual packages names are loaded properly in the selection list after issuing `guix-edit` (due to that happening in the guix repl, I presume?), the link to e.g. gnu/packages/third-party.scm is resolved to `~/.guix-profile/latest/gnu/packages/third-party.scm`, while it is actually in `$GUIX_PACKAGE_PATH/gnu/packages/third-party.scm`. 

Consequently, I cannot currently use guix.el to easily edit my out-of-tree package defintions.

I am using a git checkout of the guix tree in `~/.guix-profile/latest`, in case that might influence things.

**Comment by alezost** [comment]
Hm, `Works For Me` :confused: I use GUIX_PACKAGE_PATH myself and I've never
faced the problem you describe.

Could you evaluate (in `M-x ielm`, for example) the following:

```
(guix-package-location "<your-package-name>")
```

> I am using a git checkout of the guix tree in `~/.guix-profile/latest`, in case that might influence things.

I suppose you meant `~/.config/guix/latest`, right?  Do you use it by symlinking it to your git checkout?

Also could you show the value of GUIX_PACKAGE_PATH?
Also did you set `guix-directory` variable?

**Comment by wordempire** [comment]
`>(guix-package-location "zeal")
"gnu/packages/third-party.scm:160:2"`

You are correct regarding `~/.config/guix/latest` being symlinked to my git
checkout.

Also, without me setting anything myself:
`>guix-directory
"/home/jelle/.config/guix/latest"`

And my GUIX_PACKAGE_PATH is:
`$ echo $GUIX_PACKAGE_PATH/
/home/jelle/guix-pkgs/`

I could easily use the prefix argument and entering my GUIX_PACKAGE_PATH
when using guix-edit & co, but that is a bit of a hassle to keep track of.



2017-04-09 20:07 GMT+02:00 Alex Kost <notifications@github.com>:

**Comment by alezost** [comment]
> (guix-package-location "zeal")
> "gnu/packages/third-party.scm:160:2"

OK, so this is the problem.  For me, all packages that I have in
GUIX_PACKAGE_PATH have absolute file names, and the packages from Guix
have relative file names ("gnu/packages/...").  So I have:

    (guix-package-location "zeal")
    "/home/al/config/guix/guix-package-path/gnu/packages/third-party.scm:7:2"

That's why it works for me.  And to be honest, I don't know why you have
a relative file name.  Could you try the following guile script to check
if it also outputs the relative file name:

```scheme
(use-modules
 (guix packages)
 (guix utils)
 (gnu packages third-party))

(display (location-file (package-location zeal)))
(newline)
```

Run it like this:

    $ GUILE_LOAD_COMPILED_PATH=/home/jelle/.config/guix/latest GUILE_LOAD_PATH=/home/jelle/guix-pkgs:/home/jelle/.config/guix/latest guile <the-above-script.scm>

Does it output the same file name?

Also I wonder, what if you put your packages just in `(third-party)`
module (or whatever you'll call it), without "gnu packages" part, I mean
into `~/guix-pkgs/third-party.scm` file.  What file name do you get in
this case?

**Comment by wordempire** [comment]
Running your guile snippet, I get the correct answer:
`/home/jelle/guix-pkgs/gnu/packages/third-party.scm`

Changing my module to reside in the top level guix-pkgs folder does seem to work. I would still say something somewhere is going wrong, but this workaround is good enough for me. You can close this issue if you want, thanks for the help 👍

**Comment by alezost** [comment]
Hm, interesting.

I see that Guix uses `current-source-location` Guile procedure (in
`(guix packages)` module) to define package location.  And I wonder why
sometimes it is relative and sometimes absolute.

Anyway, it's not an Emacs-Guix issue (either Guix or Guile).

Closing, thanks.

**Comment by jellelicht** [comment]
Hey alezost,

if I am not mistaken, the current  [guix-devel ML](lists.gnu.org/archive/html/guix-devel/2017-07/msg00049.html) discussion is referring to exactly this issue as I was experiencing it.

Invoking `guix edit` in a terminal seems to be able to properly resolve any package, so maybe guix.el could do something similar?

**Comment by alezost** [comment]
Yeah, you are right.  This is now in the TODO list :smile:

**Comment by alezost** [comment]
I have finally found the time to fix it (by commit f85e1bc2666a9d668c808984757862c9aa93d269) :blush:

**Comment by jellelicht** [comment]
thanks! It makes using emacs for managing guix-related things a lot nicer ;-).
